src/core/lib/event_engine/windows/windows_listener.cc

```diff
@@ -0,0 +1,323 @@
+      port_(port),
+      listener_socket_(std::move(win_socket)),
+      listener_sockname_(hostbyname) {}
+
+absl::StatusOr<WindowsEventEngineListener::SinglePortSocketListener::
+                   PrepareListenerSocketResult>
+WindowsEventEngineListener::SinglePortSocketListener::PrepareListenerSocket(
+    SOCKET sock, const EventEngine::ResolvedAddress& addr) {
+  auto fail = [&](absl::Status error) -> absl::Status {
+    GPR_ASSERT(!error.ok());
+    auto addr_uri = ResolvedAddressToURI(addr);
+    error = grpc_error_set_int(
+        grpc_error_set_str(
+            GRPC_ERROR_CREATE_REFERENCING("Failed to prepare server socket",
+                                          &error, 1),
+            grpc_core::StatusStrProperty::kTargetAddress,
+            addr_uri.ok() ? *addr_uri : addr_uri.status().ToString()),
+        grpc_core::StatusIntProperty::kFd, static_cast<intptr_t>(sock));
+    if (sock != INVALID_SOCKET) closesocket(sock);
+    return error;
+  };
+
+  auto error = PrepareSocket(sock);
+  if (!error.ok()) return fail(error);
+  if (bind(sock, addr.address(), addr.size()) == SOCKET_ERROR) {
+    return fail(GRPC_WSA_ERROR(WSAGetLastError(), "bind"));
+  }
+  if (listen(sock, SOMAXCONN) == SOCKET_ERROR) {
+    return fail(GRPC_WSA_ERROR(WSAGetLastError(), "listen"));
+  }
+  int sockname_temp_len = sizeof(struct sockaddr_storage);
+  EventEngine::ResolvedAddress sockname_temp;
+  if (getsockname(sock, const_cast<sockaddr*>(sockname_temp.address()),
+                  &sockname_temp_len) == SOCKET_ERROR) {
+    return fail(GRPC_WSA_ERROR(WSAGetLastError(), "getsockname"));
+  }
+  sockname_temp =
+      EventEngine::ResolvedAddress(sockname_temp.address(), sockname_temp_len);
+  return PrepareListenerSocketResult{ResolvedAddressGetPort(sockname_temp),
+                                     sockname_temp};
+}
+
+// ---- WindowsEventEngineListener ----
+
+WindowsEventEngineListener::WindowsEventEngineListener(
+    IOCP* iocp, AcceptCallback accept_cb,
+    absl::AnyInvocable<void(absl::Status)> on_shutdown,
+    std::unique_ptr<MemoryAllocatorFactory> memory_allocator_factory,
+    Executor* executor, const EndpointConfig& config)
+    : iocp_(iocp),
+      accept_cb_(std::move(accept_cb)),
+      on_shutdown_(std::move(on_shutdown)),
+      memory_allocator_factory_(std::move(memory_allocator_factory)),
+      config_(config),
+      executor_(executor) {}
+
+WindowsEventEngineListener::~WindowsEventEngineListener() {
+  auto shutdown_msg =
+      absl::StrFormat("WindowsEventEngineListener::%p shutting down", this);
+  GRPC_EVENT_ENGINE_TRACE("%s", shutdown_msg.c_str());
+  // Shut down each port listener before destroying this EventEngine::Listener
+  for (auto& port_listener : port_listeners_) {
+    port_listener.reset();
+  }
+  on_shutdown_(absl::OkStatus());
+}
+
+absl::StatusOr<int> WindowsEventEngineListener::Bind(
+    const EventEngine::ResolvedAddress& addr) {
+  if (started_) {
```

Comment:
I think accesses to `started_` should either be synchronized, or this should really just be an assert. (again as discussed)

Reply:
I made it an atomic, and also asserted that `Start` is only called once. Recapping the discussion a bit: if gRPC code is simultaneously calling `Bind` and `Start`, that's improper usage. This is a bit of defensive programming, but we might legitimately have improper usage that needs to be uncovered, so it's probably for the best to not crash here quite yet.